core/src/main/java/hudson/model/Actionable.java

```diff
@@ -91,18 +95,51 @@
      */
     @Exported(name="actions")
     public final List<? extends Action> getAllActions() {
-        List<Action> _actions = new ArrayList<Action>(getActions());
-        for (TransientActionFactory<?> taf : ExtensionList.lookup(TransientActionFactory.class)) {
-            if (taf.type().isInstance(this)) {
-                try {
-                    _actions.addAll(createFor(taf));
-                } catch (Exception e) {
-                    LOGGER.log(Level.SEVERE, "Could not load actions from " + taf + " for " + this, e);
+        List<Action> _actions = getActions();
+        boolean adding = false;
+        synchronized (Actionable.class) {
+            if (factoryCache == null) {
+                @SuppressWarnings("rawtypes")
+                final ExtensionList<TransientActionFactory> allFactories = ExtensionList.lookup(TransientActionFactory.class);
+                factoryCache = CacheBuilder.newBuilder().build(new CacheLoader<Class<? extends Actionable>, Collection<? extends TransientActionFactory<?>>>() {
+                    @Override
+                    public Collection<? extends TransientActionFactory<?>> load(Class<? extends Actionable> implType) throws Exception {
+                        List<TransientActionFactory<?>> factories = new ArrayList<>();
+                        for (TransientActionFactory<?> taf : allFactories) {
+                            if (taf.type().isAssignableFrom(implType)) {
+                                factories.add(taf);
+                            }
+                        }
+                        return factories;
+                    }
+                });
+                allFactories.addListener(new ExtensionListListener() {
+                    @Override
+                    public void onChange() {
+                        factoryCache.invalidateAll();
```

Comment:
Fair.  I can see other places this strategy might be useful, so may borrow it in the future.